src/Compiler/Driver/CompilerDiagnostics.fs

```diff
@@ -2176,10 +2176,10 @@ type DiagnosticsLoggerFilteringByScopedPragmas
             if report then
                 if diagnostic.ReportAsError(diagnosticOptions, severity) then
                     diagnosticsLogger.DiagnosticSink(diagnostic, FSharpDiagnosticSeverity.Error)
-                elif diagnostic.ReportAsWarning(diagnosticOptions, severity) then
-                    diagnosticsLogger.DiagnosticSink(diagnostic, FSharpDiagnosticSeverity.Warning)
                 elif diagnostic.ReportAsInfo(diagnosticOptions, severity) then
                     diagnosticsLogger.DiagnosticSink(diagnostic, severity)
+                elif diagnostic.ReportAsWarning(diagnosticOptions, severity) then
```

Comment:
I tried lots of other approaches but it always came down to the issue, that informational diagnostics are promoted too liberally into warnings when using the `DiagnosticsLoggerFilteringByScopedPragmas` 🤷

Reply:
To strengthen the argument that the old behaviour of `DiagnosticsLoggerFilteringByScopedPragmas` was wrong, I extended the test in `Test NoWarn HashDirective` to emit diagnostics that should be of Severity `Info`, see [here](https://github.com/dotnet/fsharp/blob/b8979083b6a93823fdee72a950da4a09869f8481/src/Compiler/Checking/TypeRelations.fs#L187)

Without the elif order change to `DiagnosticsLoggerFilteringByScopedPragmas` the test fails for `useTransparentCompiler=useTransparentCompiler = false|true` because the Info Diagnostic gets promoted to a Warning.

Reply:
I remember at some point I had the same issue and thought what was expected by the tests was incorrect. Then after some adjustments to resolve another diagnostics issue it went away (i.e. started behaving like IncrementalBuilder) so I didn't pursue it further.

Could the new behavior under some circumstances "break" existing code? If not then we should probably adjust the tests.

Reply:
Okay, the tests threw me off in my understanding what `warnon` is supposed to do.
I wrongly assumed it only means to report diagnostics with severity `Info`.
But of course it really mean to propagate them to severity `Warning`, too.

The thing with the test in `ObjInference.fs` is that the call to `typecheck` is not yet doing that severity propagation, that only happens when we call `compile`. I just wasn't aware of that difference.

I adjusted the test to check the diagnostics after the propagation (so after compile).
If I'm not again totally confused the behaviour of the TransparentCompiler is better than that of the old BackgroundCompiler as the propagation happens already during TC which seems to be more in line with what the user expects when using a `warnon` flag. I hope that makes sense 😬

Reply:
Yeah I think that makes sense 👍